File: src/main/java/jp/kentan/minecraft/neko_event/config/provider/GachaConfigProvider.java
package jp.kentan.minecraft.neko_event.config.provider;

import jp.kentan.minecraft.neko_event.config.ConfigManager;
import jp.kentan.minecraft.neko_event.config.listener.ConfigListener;
import jp.kentan.minecraft.neko_event.gacha.model.Gacha;
import jp.kentan.minecraft.neko_event.util.Log;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;


public class GachaConfigProvider {

    private static File sConfigFile;

    private static ConfigListener<Gacha> sListener;

    public static void setup(File dataFolder){
        sConfigFile = new File(dataFolder + File.separator + "gacha.yml");

        ConfigManager.createIfNeed(sConfigFile);
    }

    public static void bindListener(ConfigListener<Gacha> listener){
        sListener = listener;
    }

    public static void load(){
        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            final Map<String, Gacha> gachaMap = new HashMap<>();

            if(!config.isConfigurationSection("Gacha")){
                Log.warn("Gacha section not found.");
                return;
            }

            final Set<String> gachaIdSet = config.getConfigurationSection("Gacha").getKeys(false);

            gachaIdSet.forEach(id -> {
                final String path = "Gacha." + id;

                Gacha gacha = new Gacha(id,
                        config.getString(path + ".name", "ガチャ"),
                        config.getString(path + ".getMsg"),
                        config.getString(path + ".missMsg"),
                        config.getString(path + ".broadcastMsg"),
                        config.getBoolean(path + ".enableEffect", true),
                        config.getBoolean(path + ".requireVoteTicket", false)
                );

                final String componentsPath = "Gacha." + id + ".components";
                final Set<String> components = config.getConfigurationSection(componentsPath).getKeys(false);
                components.forEach(index -> {
                    final String indexPath = componentsPath + "." + index;

                    gacha.add(
                            config.getString(indexPath + ".name"),
                            config.getDouble(indexPath + ".probability"),
                            config.getStringList(indexPath + ".commands"));
                });

                gacha.normalizeIfNeed();
                gachaMap.put(id, gacha);
            });

            Log.info(gachaMap.size() + "個のガチャを読み込みました.");

            sListener.onUpdate(gachaMap);
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
        }
    }
}

File: src/main/java/jp/kentan/minecraft/nekoevent/config/provider/GachaConfigProvider.kt
package jp.kentan.minecraft.nekoevent.config.provider

import jp.kentan.minecraft.nekoevent.component.model.Gacha
import jp.kentan.minecraft.nekoevent.config.ConfigUpdateListener
import jp.kentan.minecraft.nekoevent.util.Log
import jp.kentan.minecraft.nekoevent.util.formatColorCode
import org.bukkit.configuration.file.YamlConfiguration
import java.io.File

class GachaConfigProvider(dataFolder: File) : BaseConfigProvider(dataFolder, "gacha.yml") {

    var listener: ConfigUpdateListener<Gacha>? = null

    fun load() {
        try {
            val config = YamlConfiguration.loadConfiguration(configFile)

            if (!config.isConfigurationSection("Gacha")) {
                Log.warn("Gachaセクションがありません.")
                return
            }

            val gachaIdSet = config.getConfigurationSection("Gacha").getKeys(false)

            val gachaMap: Map<String, Gacha> = gachaIdSet.mapNotNull { id ->
                val path = "Gacha.$id"

                // Gacha Component
                val componentsPath = "Gacha.$id.components"
                val componentIdSet = config.getConfigurationSection(componentsPath).getKeys(false)
                val componentList = componentIdSet.map { index ->
                    val indexPath = "$componentsPath.$index"

                    return@map Gacha.Component(
                            config.getString("$indexPath.name").formatColorCode(),
                            config.getDouble("$indexPath.probability"),
                            config.getStringList("$indexPath.commands")
                    )
                }.toMutableList()

                if (!Gacha.Component.normalize(componentList)) {
                    Log.error("ガチャ($id)の正規化に失敗しました.")
                    return@mapNotNull null
                }

                // Gacha
                return@mapNotNull id to
                    Gacha(
                        id,
                        config.getString("$path.name", "ガチャ"),
                        componentList,
                        config.getString("$path.winMessage")?.formatColorCode(),
                        config.getString("$path.loseMessage")?.formatColorCode(),
                        config.getString("$path.broadcastMessage")?.formatColorCode(),
                        config.getBoolean("$path.enabledEffect", true)
                    )
            }.toMap()

            listener?.onConfigUpdate(gachaMap)
        } catch (e: Exception) {
            e.printStackTrace()
            Log.error(e)
        }
    }
}